### PR TITLE
[FIX] *_expense: update approval_state and forbids unrealistic case when refusing expense sheet

### PR DESCRIPTION
**Pull request by daiduongnguyen-odoo**
*= hr_expense, project_sale_expense

-STEP TO REPRODUCE: Create an expense sheet , submit then approve it as
well, go to Pg admin or update the view manually to see the field
`approval_state`. The `approval_state` is now approved.
After that try to refuse the sheet, `approval_state` still approved
while it should be consider `cancel` (Refuse)

-Solution is to forbids unrealistic case where we shouldn't allow user
to cancel when it linked to a journal entry and also update `approval_state` as well -> Therefore some test in
`test_project_profitability` need to adapt

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by duong77476-viindoo** [comment]
Hello @JulienAlardot Can you take a look when you have free time ? Thank you

**Review by casp-odoo — COMMENTED** [review]
Why you would have this specific test case fail this way is beyond me. Try rebasing ? In any case, you'll need to get it to pass. 
I don't see objections, but I don't know much about this process. Maybe @jbw-odoo knows more.

**Comment by JulienAlardot** [comment]
> Why you would have this specific test case fail this way is beyond me. Try rebasing ? In any case, you'll need to get it to pass. I don't see objections, but I don't know much about this process. Maybe @jbw-odoo knows more.

Long story short, the domain of _get_expenses_profitability is bad, and this PR now shows how much. I will make a PR to fix it. The project_same_expense & project_hr_expense modules need a refactor

**Comment by duong77476-viindoo** [comment]
@JulienAlardot Yes, i assume the error at `_get_expenses_profitability_items` , so can my PR consider merge ?
cc @casp-odoo @jbw-odoo

**Comment by JulienAlardot** [comment]
We can't merge it unless the test & base function are fixed

**Comment by duong77476-viindoo** [comment]
> We can't merge it unless the test & base function are fixed

@JulienAlardot 
Okay, i'll try fix or i 'll wait for you since you said it need refractor.

File: addons/hr_expense/models/hr_expense_sheet.py
Comment (duong77476-viindoo) [inline addons/hr_expense/models/hr_expense_sheet.py:661]: @JulienAlardot I know why, the `compute_state` of expense sheet will trigger again when we change `approve_state` and of course the state of `expense` is computed base on `expense sheet` so i write like this, everything is ok and the failed test we 've seen earlier is pass, not sure for the other. If you ok with this approach i'll leave a code comment inside

Comment (JulienAlardot) [reply addons/hr_expense/models/hr_expense_sheet.py:661]: Please revert to the previous one, this won't fix anything. Calling write twice is bad practice, in theory we could just reset the approval state and the state would be recomputed but IMO we can already specify it here, it's less computation

**Comment by JulienAlardot** [comment]
@duong77476-viindoo Hello, please consider these changes
https://github.com/odoo-dev/odoo/commit/04af3d8fbca852f13df40d294c2a7d7739045855

**Comment by duong77476-viindoo** [comment]
@JulienAlardot seem very good sir, running automation test on some expense module pass. Thank you very much

File: addons/hr_expense/i18n/hr_expense.pot
Comment (duong77476-viindoo) [inline addons/hr_expense/i18n/hr_expense.pot:2309]: I export the term only

File: addons/hr_expense/models/hr_expense_sheet.py
Comment (duong77476-viindoo) [inline addons/hr_expense/models/hr_expense_sheet.py:662]: I think it should be `self.write({'state': 'cancel'})` unlike yours sir

File: addons/hr_expense/models/hr_expense_sheet.py
Comment (duong77476-viindoo) [reply addons/hr_expense/models/hr_expense_sheet.py]: Ah i'm just being dump sr

File: addons/hr_expense/models/hr_expense_sheet.py
Comment (JulienAlardot) [reply addons/hr_expense/models/hr_expense_sheet.py:662]: calling .write on a single field is a bad practice that hinders readability
We could update update both fields at the same time, but I think it's better to use the compute in case the behaviour changes in the future

Comment (duong77476-viindoo) [reply addons/hr_expense/models/hr_expense_sheet.py]: okay then

**Comment by duong77476-viindoo** [comment]
Hello @JulienAlardot You still working on this sir ?

File: addons/hr_expense/models/hr_expense_sheet.py
Comment (JulienAlardot) [inline addons/hr_expense/models/hr_expense_sheet.py:660]:
```suggestion
        if any(self.account_move_ids):  # Todo: in 17.3+, edit it to allow draft entries"
```

**Review by JulienAlardot — CHANGES_REQUESTED** [review]
Small thing I didn't notice in the patch I proposed.
After that change it's good for me to merge
Sorry for the wait and thank you :)

File: addons/hr_expense/models/hr_expense_sheet.py
Comment (duong77476-viindoo) [reply addons/hr_expense/models/hr_expense_sheet.py]: @JulienAlardot Better wait than never, thanks for the review

**Comment by JulienAlardot** [comment]
LGTM @antoine162

**Review by JulienAlardot — APPROVED** [review]

**Review by antoine162 — APPROVED** [review]
Looks great, thanks so much @duong77476-viindoo @JulienAlardot 
@robodoo r+

**Pull request closed without merging**
